click/types.py

```diff
@@ -479,10 +481,51 @@ def convert(self, value, param, ctx):
                     self.path_type,
                     filename_to_ui(value)
                 ), param, ctx)
+            if self.executable and not os.access(value, os.X_OK):
+                self.fail('%s "%s" is not executable.' % (
+                    self.path_type,
+                    filename_to_ui(value)
+                ), param, ctx)
 
         return self.coerce_path_result(rv)
 
 
+class Executable(Path):
+    """The executable type is similar to the :class:`Path` type but checks
+    for an executable file.  A file without path information will be looked
+    up in $PATH.
+
+    For more information see :ref:`tuple-type`.
+
+    :param absolute_path: use the absolute (but not necessarily resolved path).
+    :param resolve_path: if this is true, then the path is fully resolved
+                         before the value is passed onwards.  This means
+                         that it is absolute and symlinks are resolved.  It
+                         will not expand a tilde-prefix, as this is
+                         supposed to be done by the shell only.
+    :param writable: if true, a writable check is performed.
+    :param readable: if true, a readable check is performed.
+    """
+    def __init__(self, absolute_path=True, resolve_path=False, writable=False,
+                 readable=False):
+        self.absolute_path = absolute_path
+        super(Executable, self).__init__(
+            exists=True, file_okay=True, dir_okay=False,
+            writable=writable, readable=readable, executable=True,
+            resolve_path=False, allow_dash=False, path_type=None)
+
+    def convert(self, value, param, ctx):
+        from shutil import which
```

Comment:
AFAIK `shutil.which()` only exists in recent Python 3 versions.

Reply:
yes, would need to be vendored, see https://github.com/pallets/click/issues/856.

Reply:
Should I do this?
I would add it to click/_compat.py, as `which`?!